bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/ProgramDiff.scala

```diff
@@ -0,0 +1,125 @@
+    * is still in the remote program or (b) has different version information
+    * than the remote program.
+    *
+    * - `Present` diffs for any node in the local program with different
+    * version information than its remote counterpart.
+    *
+    * - `Present` diffs for all nodes in an observation that have at least one
+    * contained node with a difference in version data.  In other words, any
+    * difference in an observation will bring `Present` diffs for the entire
+    * observation.
+    *
+    * - `Present` diffs for all ancestor nodes of a node with different version
+    * data.  For example, an edited note in a group will pull in the group and
+    * the program ancestors.
+    *
+    * Conversely, the comparison must not
+    *
+    * - include `Present` diffs for any node which has the same version
+    * information as the remote node and which has no descendants with different
+    * version data
+    *
+    * - include `Missing` diffs for nodes deleted both locally and remotely and
+    * that have the same version data
+    *
+    * @param local local program from which differences are to be extracted
+    *
+    * @param remoteVm `VersionMap` of the remote instance of the program
+    *
+    * @param remoteRemoved `Set` of all node keys that are not present in the
+    *                     remote instance of the program
+    *
+    * @return super-set of `Diff`s between the local and remote program
+    *         instances
+    */
+  def compare(local: ISPProgram, remoteVm: VersionMap, remoteRemoved: Set[SPNodeKey]): List[Diff] = {
+    def versionDiffers(k: SPNodeKey): Boolean =
+      remoteVm.get(k).forall(_ =/= local.getVersions(k))
+
+    // locally present node differs from the remote version
+    def presentDiffers(k: SPNodeKey): Boolean =
+      versionDiffers(k) || remoteRemoved.contains(k)
+
+    // locally missing node differs from the remote version
+    def missingDiffers(k: SPNodeKey): Boolean =
+      versionDiffers(k) || !remoteRemoved.contains(k)
+
+    def nodeDiffers(n: ISPNode): Boolean = presentDiffers(n.key)
+
+    def diffNode(n: ISPNode): Option[Diff] = nodeDiffers(n) option present(n)
+
+    // Present differences in in-use nodes rooted at r.
+    def presentDiffs(r: ISPNode): List[Diff] =
+      r match {
+        case o: ISPObservation =>
+          // Observations are atomic.  If anything differs at all in either
+          // version copy the entire observation.
+          if (o.exists(nodeDiffers)) Diff.tree(o) else List.empty
+
+        case _                 =>
+          val childDiffs = (List.empty[Diff]/:r.children) { (ns, child) =>
+            presentDiffs(child) ++ ns
+          }
+
+          // If there is even one descendant that differs, include this node
+          // in the results.  Otherwise, only include it if it differs.
+          if (childDiffs.isEmpty) diffNode(r).toList
+          else present(r) :: childDiffs
+      }
```

Comment:
I think I would find this a little easier to follow if it were a single expression, although as you noted your fold might be more efficient in the end.

```scala
case _ =>
  // If there is even one descendant that differs, include this node
  // in the results.  Otherwise, only include it if it differs.
  r.children.flatMap(presentDiffs) match {
    case Nil => diffNode(r).toList
    case xs  => present(r) :: xs
  }
```